module/zfs/zfs_replay.c

```diff
@@ -585,6 +594,9 @@ zfs_replay_remove(void *arg1, void *arg2, boolean_t byteswap)
 		error = SET_ERROR(ENOTSUP);
 	}
 
+#ifdef __FreeBSD__
+	VOP_UNLOCK(ZTOV(dzp), 0);
+#endif
 	zrele(dzp);
```

Comment:
How does `VOP_UNLOCK()` locks work compared to `zrele()` locks, so that you need both in FreeBSD?

Reply:
It's my understanding there are some checks in the FreeBSD which require the lock.  We could add a wrapper, but having it do nothing on other platforms would probably be confusing.  Better ideas welcome, since this is ugly.

Reply:
Manpage says `These calls are used to serialize access to the file system, such as to
     prevent two writes	to the same file from happening	at the same time`
Makes me wonder how I am protecting against that on osx (especially since locking vnodes is not something I'm allowed to call). Although the internals to `vn_rdwr()` appear to do the same locks, if not taken: https://github.com/freebsd/freebsd/blob/bcfa67ab8b8212a8a15763e8a6855a272b8fde22/sys/kern/vfs_vnops.c#L554   But I was just curious about the difference

Reply:
@lundman The problem is that FreeBSD currently does all vnode locking in the VFS layer. So the underlying primitives can't do any locking themselves on FreeBSD. I could have an additional duplicate wrapper that duplicated the locking here, but in the next year we plan on fixing VFS for non rubbish consumers.